jq255.md

```diff
@@ -0,0 +1,1412 @@
+
+`is_square(x)` can be implemented as a simple wrapper around `sqrt(x)`,
+though faster implementations are possible with the [Legendre/Jacobi
+symbol]. A fast, constant-time implementation of that operation can be
+obtained as a straightforward derivative of the [optimized binary
+GCD]; on some small microcontrollers, this `is_square()`
+implementation can be about six times faster than the full square root
+computation.
+
+### Affine Formulas
+
+We recall here the affine formulas applicable to points on double-odd
+Jacobi quartic curves. Implementations are expected to use the extended
+representation described later on in this document; the affine formulas
+are potentially useful if some alternate point representations are
+explored.
+
+For any point `P = (e, u)` on the curve, the point `P+N` (i.e. the other
+representant of the same group element) is `P+N = (-e, -u)`. The
+*opposite* of the point `P` is `-P = (e, -u)`. The neutral element of
+the group is `{ I, N }` (i.e. `I = (1, 0)` and `N = (-1, 0)` are the two
+representants of the group neutral).
+
+Let `P1 = (e1, u1)` and `P2 = (e2, u2)` be two points on the curve. Their
+sum `P3 = P1 + P2 = (e3, u3)` can be computed as follows:
+
+```
+     (1 + b'*u1^2*u2^2)*(e1*e2 + a'*u1*u2) + 2*b'*u1*u2*(u1^2 + u2^2)
+e3 = ----------------------------------------------------------------
+                          (1 - b'*u1^2*u2^2)^2
+
+      e1*u2 + e2*u1
+u3 = ----------------
+     1 - b'*u1^2*u2^2
+```
+
+These formulas are *complete*: they work on all pairs of curve points,
+including the points `I` and `N` (that represent the neutral element).
+
+The Jacobi quartic curve `e^2 = b'*u^4 + a'*u^2 + 1` is birationally
+equivalent to the curve with Weierstraß equation `y^2 = x*(x^2 + a*x + b)`
+with the following mappings (defined for all points excepts `I` and `N`):
+
+```
+x = (e + 1 - a*u^2)/(2*u^2)
+y = x/u
+
+e = (x^2 - b)/(x^2 + a*x + b)
+u = x/y
+```
+
+On the Weierstraß curve, `N` has coordinates `(0, 0)`, and `I` has no
+defined coordinates (it is then truly "at infinity"). The Weierstraß
+equation can then be converted to a *short Weierstraß equation*
+`Y^2 = X^3 + A*X + B` by using the following:
+```
+A = (3*b - a^2)/3
+B = a*(2*a^2 - 9*b)/27
+X = x + a/3
+Y = y
+```
+
+Converting to Weierstraß coordinates and back is not required for
+implementing the jq255e and jq255s groups; such coordinates have
+furthermore special cases for the points `I` and `N`. These conversions
+may be useful in some specific contexts, e.g. to leverage preexisting
+generic implementations of (short) Weierstraß curves, if only for
+testing purposes.
+
+### Extended Coordinates
+
+In order to reduce the number of field element divisions (which are
+usually quite expensive in software implementations, relative to
+additions and multiplications), the following internal representation is
+recommended: a point `P = (e, u)` is represented by the quadruplet
+`(E:Z:U:T)` of field elements, with the following rules:
```

Comment:
This is mildly surprising, as I am used to X:Y:Z:T coordinates. Is the inversion of U and Z intentional?

Reply:
Yes, with the inversion it makes it sound like a French mild expletive (equivalent to something like "oh darn", nothing dire).